### PR TITLE
Bugfix: only track UTXO modification after lookup

### PR DESCRIPTION
**Pull request by sipa**
Otherwise, if CCoinsViewCache::ModifyCoins throws an exception in between setting hasModifier and constructing the CCoinsModifier, the cache ends up in an inconsistent state, resulting in an assert failure in the next modification.

Bug discovered by @laanwj.

### CODE REVIEW

**Comment by laanwj** [comment]
Tested ACK (e.g. https://github.com/laanwj/bitcoin/tree/2014_12_leveldb_crash_fuzzing )

**Comment by sipa** [comment]
Added a commit to catch and shutdown gracefully in case of LevelDB errors during flush.

**Comment by sipa** [comment]
This is not sufficient (as testing by @gmaxwell shows), there are many cases where reads from the database can be triggered and aren't caught.

**Comment by laanwj** [comment]
Yes, leveldb exceptions an escape from CCoinsViewDB in many other places as just the flush, and if they happen in the network code they are caught in the top-level handler, be logged, and the process will happily continue.

However, as this is strictly an improvement we should still merge this?

**Comment by sipa** [comment]
@laanwj Yes, I think it's definitely an improvement.

**Comment by gmaxwell** [comment]
Tested ACK.  (As pieter notes, we still have things to fit in this respect, but better is better)

File: src/main.cpp [hunk 1]
```diff
@@ -1807,6 +1807,7 @@ enum FlushStateMode {
 bool static FlushStateToDisk(CValidationState &state, FlushStateMode mode) {
     LOCK(cs_main);
     static int64_t nLastWrite = 0;
+    try {
```
Comment (Diapolo) [inline src/main.cpp]: @sipa Why did you choose to leave out the indentation? Our code get's pretty ugly again over time? Could we finally let the clang-format tool let do it's job and we also had a remove orphan spaces and tabs script if I remember correctly...

**Pull request merged**